### ISSUE TITLE
API performance issue 

### ISSUE THREAD

**Issue by neilpellowe** [reapit/foundations#8548]
Connells have raised an issue with the API performance when testing their GEM integration in their LIVE environment. Details of the issue initially sent to Nick T by Daniel @ Connells below.

"We are calling the URL as follow
 
[https://platform.reapit.cloud/appointments/?pageNumber=1&pageSize=100&negotiatorId=ANX0&officeId=MLB&start=2023-02-13T00%3A00%3A00.0000000Z&end=2023-03-04T00%3A00%3A00.0000000Z](https://protect-eu.mimecast.com/s/iOXNCyP4WHrRMzghQZ3xm?domain=platform.reapit.cloud)
 
As you can see the “start” and “end” parameters are correctly added to the URI.
 
I’ve ran this test a few times against the production Reapit API and the most recent “x-amzn-RequestId” Reapit need to track it their side is:
 
20abc0de-87a4-4dfd-86e2-dedd88463e14
 
 
Basically, when we do the call on the sandbox, it takes 300ms where as we’re seeing times of 5 seconds."

**Comment by HollyJoyPhillips** [comment]
App ID: 8b9ae86c-2a71-40df-9d50-14f83b3518b7 
Customer ID: CWC

**Comment by plittlewood-rpt** [comment]
We’ve had an initial look at this and without some fairly heavy refactoring the query is probably as light as it can be, with the issue stemming from having to look up recurring appointments as part of the query (which we do when certain query string parameters are set) that fall within the requested date range.

An immediate observation from an AC perspective is that the diary table needs some housekeeping. 50% of it is over a year old and could/should be moved to the archive which I would expect to result in a fairly dramatic improvement.

We are making some changes our end which we hope to be in production tomorrow morning which should see some improvement. Once we've deployed that we will be in a better position to review again.

**Comment by plittlewood-rpt** [comment]
Two performance tweaks have gone into the appointments service with the latter just hitting production. This seems to have done the trick on this database. Please can we get some feedback from the customer. I've reached out to Nick so please touch base with him so we don't all go back to them. Thanks

**Comment by HollyJoyPhillips** [comment]
I'll message the customer, thank you

**Comment by HollyJoyPhillips** [comment]
"We have tested it, it’s 4/5 times faster which is great but… it’s bringing back data greater than what we asked for in the query. Are you or your team able to advise why?"

I've asked the customer to reply/subscribe to the ticket direct going forward.

**Comment by plittlewood-rpt** [comment]
This could be because the recurring appointments aren't being interpreted properly. When you pass start/end dates in your query the API will (by design) return any recurring appointments whose _original_ appointment date is outside that range. This has caught people out before, so let's wait for more information before proceeding further

**Comment by danielalmeidaconnellsgroup** [comment]
Hi All

Appointments get for Frank Manning took 641
Earliest Appointment returned 13/02/2023 09:00:00; from 16/02/2023 00:00:00 <-- shown  first DT is what API returned as the earliest appointment, Second value is the from date in the request so 1st date shouldn't be less than 2nd date on that line. (this is from Sandbox so is ok)


Appointments get for Abul Niloy took 10797
Earliest Appointment returned 17/08/2022 16:00:00; from 16/02/2023 00:00:00
Appointments get for Admin Centre took 2359
Earliest Appointment returned ; from 16/02/2023 00:00:00
Appointments get for Henry Slade took 2641
Earliest Appointment returned 05/01/2018 09:00:00; from 16/02/2023 00:00:00
Appointments get for Jean-Paul Francis took 2323
Earliest Appointment returned ; from 16/02/2023 00:00:00
Appointments get for Matthew Ward took 2483
Earliest Appointment returned 10/08/2021 10:00:00; from 16/02/2023 00:00:00
Appointments get for Nicola Rudsits took 3024
Earliest Appointment returned 15/07/2022 09:30:00; from 16/02/2023 00:00:00
Appointments get for Richard Olliffe took 2275
Earliest Appointment returned ; from 16/02/2023 00:00:00
Appointments get for Scott Burman took 2235
Earliest Appointment returned 17/02/2023 12:30:00; from 16/02/2023 00:00:00
Appointments get for Scott Garlick took 2724
Earliest Appointment returned 09/06/2022 01:00:00; from 16/02/2023 00:00:00


That's Taylors LB again, all results except Rich Ollife returned appointments earlier than the from date

We haven't checked for end date being exceeded but that also might be off. Times are in milliseconds.

**Comment by plittlewood-rpt** [comment]
Hi @danielalmeidaconnellsgroup please can you check your results to see if the ones whose dates fall outside the request dates are actually recurring appointments. This has caught a few people out. Appointments that are marked as recurring where they would fall in the date range specified will still be returned, it is down to you as the consumer to interpret the recurrence information. If you can let me know your thoughts once you've checked this I'd be grateful. Thanks

**Comment by danielalmeidaconnellsgroup** [comment]
Hello, we will check but... we need a way to not bring back recurring appointments. Instead we just want to see diary availability for a set number of days. Is there no parameter to exclude these which we can pass on the API?

**Comment by plittlewood-rpt** [comment]
Hi @danielalmeidaconnellsgroup there is currently no way to exclude recurring appointments, however I'd recommend NOT excluding these, as you'll be omitting records for appointments where there isn't actually a space in that user's diary. You need the recurring appointments to be able to have a full picture of the diary of that user

**Comment by danielalmeidaconnellsgroup** [comment]
Agreed - I'm not saying we  don't want the appointments during that time, you're completely right - we just don't appointments which are outside of the time frame in question.

**Comment by plittlewood-rpt** [comment]
Have a look at how the recurring appointments are surfaced. They will appear to be outside the date range but that is because the original appointment is returned (as we don't explicitly store copies for each recurrence). Evaluating the recurrence data in that appointment will tell you where the recurring ones would fall inside the data range you've queried if that makes sense. Let me know if you need me to clarify anything

**Comment by danielalmeidaconnellsgroup** [comment]
Thanks for your help. We have reviewed made some changes which have improved performance enough to allow us to progress. We'll pick up on how we handle this long term as part of our ongoing discussions. Thank you for all your help and for the changes you did to move it from 2.5mins to 30s.

**Comment by plittlewood-rpt** [comment]
Thanks @danielalmeidaconnellsgroup are you happy for this to be closed. Performance is definitely high on our agenda but as you'll appreciate there's no overnight solution on larger datasets so it'll be an iterative process
